### DISCUSSION TITLE
supporting arabic

### DISCUSSION THREAD

**Discussion by king-dahmanus** [RHVoice/RHVoice#436] [Q&A] [answered]
Hey rhvoice devs and contributors. Can you tell me how I can make a dictionary for arabic, knwing that I'm not a developer? Does it require c++ programming? Can you site me to the documentation on how to do that? Thanks

**Comment by zstanecic** [comment]
Hi!

For making a language, you need to know foma, and the data language structure. And you need to know the linguistic aspect. 

You will need to create also the labelling and phonemes. You have for now the Albanian language as an example,  of the data only language. When it comes to foma, there is a foma regex reference. You will google for it.

 

 

From: king-dahmanus ***@***.***> 
Sent: Wednesday, January 12, 2022 12:55 PM
To: RHVoice/RHVoice ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [RHVoice/RHVoice] supporting arabic (Issue #435)

 

Hey rhvoice devs and contributors. Can you tell me how I can make a dictionary for arabic, knwing that I'm not a developer? Does it require c++ programming? Can you site me to the documentation on how to do that? Thanks

—
Reply to this email directly, view it on GitHub <https://github.com/RHVoice/RHVoice/issues/435> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACVCDEY2C47NEEMGIE5OCYDUVVTY3ANCNFSM5LYY5UXA> .
Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>  or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub> .

**Comment by king-dahmanus** [comment]
where do I find the data only language files? I mean which repo are they in? Thanks

**Comment by zstanecic** [comment]
These are found here, in the repo.

Paths are the following:

Binary:

/data/languages/Albanian

Sources: /src/scripts/Albanian

Best,

Zvonimir

**Comment by king-dahmanus** [comment]
hhthanks, I'll try to learn from that. But wouldn't I find a problem supporting a complicated patterned language like arabic? Is there a possibility of adding a wordlist/dictionary and also the rules

**Comment by zstanecic** [comment, marked as answer]
Yes, you can do this using finite state transducers.

**Comment by king-dahmanus** [comment]
so can I learn them from the existing languages?

**Comment by zstanecic** [comment]
You can see the examples and see, what you need to do for Arabic.

**Comment by king-dahmanus** [comment]
ok gotcha. If I wanna test the changes from source and or compile them as a prtable commandline test do I need a linux system? Cause I don't have that

**Comment by zstanecic** [comment]
Wsl2 on windows 10 is enough. 

Install on the Debian or ubuntu

Sudo apt install foma foma-bin

**Comment by king-dahmanus** [comment]
so I only need to install that and somehow build it for windows on wsl?

**Comment by louderpages** [comment]
Marhaban!  I hope I can be of some help.  

Arabic creates a special problem for TTS.  Unless you intend to read only text with diacritics added, the "tashkeel", then you need another process, which RHVoice does not support.

Mostly we hope to write rules which map written symbols, "graphemes" into phonetic symbols "phonemes". DIctionaries are needed only for words where the language is not purely phonetic, and where syllable stress is not predictable. In those cases our rules are broken, and we use the dictionary for special cases.

When Arabic is written with the tashkeel, you will need no dictionary.  It is a very regularly phonetic language.  Well, at least Modern Standard Arabic.  When it is written, as it almost always is, without the tashkeel, then not even a dictionary will tell RHVoice how to pronounce the word.  

The pronunciation of most Arabic words, without tashkeel, has to be decided from the meaning of the sentence or the grammar.  This is a kind of artificial intelligence task.  There are software diacritizers out there which do a good job, and they could be used as s front end to RHVoice.   If you can get a diacritizer to read Arabic text, then give its output to RHVoice,  RHVoice can  then easily turn that text into phonemes.

We, here at Louder Pages, have some past experience with Arabic, so if you are getting serious and need some collaboration, please let us know.

Kind regards,
Louder Pages

**Comment by king-dahmanus** [comment]
hey louder pages, I do want to collaborate, this is very exciting! I know of a diacritizer called mishkal which does a good job of predicting how the words are pronounced, it's actually a significant improvement if it could be managed in rhvoice. I hope that we can pull this off, and somehow even improve mishkal for better output. There's also shakkala and shakkelha, shakkelha ignores diacritics if you type them, but it does a really really good job of synthesizing. Can we somehow talk via instant messaging, I.e discord etc so we can easily pull this off? Knowing that I'm not a dev and I have no experiences with fona what so ever

**Comment by king-dahmanus** [comment]
not synthesizing, but rather diacritizing I ment

**Comment by king-dahmanus** [comment]
also a bit offtopic, but I don't know how to learn the syntax of fona. Can you provide me with a documentation which teaches fona and allows me to learn the whole syntax, regexs and everything else I need for an optimal tts?

**Comment by king-dahmanus** [comment]
and also how to build voices from an existing dictionary, like I wanna build my own english voice, which documentation/where can I find the documentation? I couldn't find it in the doc folder

**Comment by zstanecic** [comment]
Hi!

For building existing voices, there is an article on the RHVoice wiki.

**Comment by king-dahmanus** [comment]
ok thanks I'll check it out

**Comment by king-dahmanus** [comment]
hey, I've tried configuring the speech_tools, but at the end it says: configure: error: C++ compiler cannot create executables. What are the packedges I should install for compiling all things related to tts creation? I'm asking because I'm a completely unlinuxed person who's just beginning to launch around in the field of linux terminals

**Comment by louderpages** [comment]
Hi,

I want to make sure of a couple of points about your understanding of RHVoice.

1.  Even a skilled and experienced developer is going to find the foma work challenging. 

2.  I may be wrong, but from what you write you may be thinking the RHvoice technology is like e-speak, where the speech is produced from models of human formants, etc..  RHVoice uses HTS which involves machine learning of pronunciation from recorded human speech. A lot of recorded speech, and specially designed speech.   So, there is a whole other set of techniques to be able to design a set of utterances, get them recorded, processed. marked-up and then holding hands with the machine learning algorithm.

3. This requires huge dedication for quite a lot of time.  Once someone learns to do this, it would be my hope that those skills are applied to more than one language.  (I hope you are listening our Polish voice developers! :). My congratulations and admiration to you!)

So, I hope that puts the scale of the project into perspective.  From the outside it can seem like it is almost a configuration task - you load up the software, define various  parameters for a language, then press a button. :)

Which brings me to this question:  What is your motivation, the driving force for you?   Why do you want to build an Arabic voice in RHVoice?   This is key to me in determining what kind of collaboration we could enter into.  We have limited resources, and we are very professional and perfectionists.   Perhaps this is not the place to discuss such things, in which case please email me at ***@***.***

Best wiehes,
- Louder Pages

**Comment by king-dahmanus** [comment]
the driving force for me is: I want to make a free responsive reasonable arabic voice for the arab blind community to negate the fact that we absolutely have to buy things or get them illegally, for me, my friends and all the blind arabs out there. I want them and us to have a legit voice which can read there texts in a good way while being free. All the usable voices on windows are payed, and people like me can't afford them. So I want to make a voice which is free for them to use. That is my goal

**Comment by king-dahmanus** [comment]
hhhI also want to tell you somethings about myself to maybe ditermin for you if I deserve your collaboration or not:
1, I'm not experienced with linux
2, I've worked with espeak's phonemes before but the fact that I don't know how to opperate linux made me unable to test my changes
3 I always wanted to be a programmer, but my low end hardware restricts me from doing so.
4 I can't afford *any* kind of digital money.

**Comment by louderpages** [comment]
So,, I will guess that your motivation is the challenge and fun of building a voice, and the expectation that it will be useful for many people.  Fair enough.  But I think you have some linux training to do first.

We have so many languages wanting attention that we want to know what is wrong with existing voices for any language requested, becuse we want to concentrate on those languages with the most need.

Also we need to hear from many users and any agencies which support them.

If Arabic does become a priority for Louder Pages, perhaps you would be available to collaborate on the language and phonetics testing side?

Kind regards,
- LP

**Comment by king-dahmanus** [comment]
I'd of course be available for testing, it will be my pleasure! Oh yeah, I am learning linux's command line basics so it might be useful in the future. I've tried compiling speech_tools and festival but it says that the c compiler can't create executables, even though gcc is installed. What other packeges do I need for it? I want to make my own english voice just to get familiar with the usage of the building process. Thanks

**Comment by grzezlo** [comment]
Thank you @louderpages for your appreciation, I'm myself even greater impressed by your work and understanding of RHVoice internals, which even after few months of digging and experimenting are mysterious for me.
Thank you also for some comments in the Macedonian foma, which helped me to understand few things better.
No need to say, that the work of @Olga-Yakovleva who understood internals of the HTS and other weakly-documented libraries and as a result gave us RHVoice synthesizer is even more impressive for me.
@king-dahmanus I understand your motivation, but as was said, the thing of adding new language is not trivial; I wrote my first regular expressions about 23 years ago, in the PERL programming language, and I'm working on software development since this time, and still for me the process of defining new language for RHVoice is not simple.
Even the process of understanding source code written by other people is some advanced challenge.
It's why I wrote the tutorial about creating new voice for RHVoice, to be sure that i'll not forget some important details if i'd like to return to this subject after year or so.
If you don't have the practice in Linux, and software development, the solution would be to find someone motivated, e.g. some blind software developer from your country, who would know the Arabic language specifics and could understand and follow the technical processes of creating new voices, editing xml, compiling under Linux and so on.
If you don't know any blind or sighted software developer, try to find other blind users from your country, interested in the subject; having some group of interested people, increases chances of success: maybe somebody will know somebody, who will be able to do this or that part of the process.
As I understand well, two Macedonian people who started the Kiko project, are not software developers, but somehow managed to convince their government, that preparing professional opensource voice for their language is better and cheaper idea than financing paid speech synthesizers for the blind, and they hired Louderpages to quickly and efficiently do all the work.
Few months ago, with my coleagues we talked about going similar path, to hire people from louderpages, but the chances of giving some government money on it was minimal, mainly due to lack of idea how to get in touch with the government people.
So, we give up the idea, but some time later I talked with my coleague @zstanecic. 
He's studying Polish language on the university, so is not a software developer, but it's the man who convinced me to write again some email message here and there, to try some things one more time, and also he has some important pieces of knowledge i'd not have idea where to find, e.g. about the FOMA reference. And the last but not least, the patience to thoroughly test new things, and report found problems.
If i'd try to implement Polish language for RHVoice myself, i'll for sure give it up in frustration about few thousands times.
Having somebody else in the team, is crucial.
So we have pretty well-working Polish language implementation, are doing some initial home recordings for new voices preparation, and it's the time when my other coleague enters the game: his friend has a professional recording studio, and his wife with good voice agrees to record herself for our project.
Few days and one night  later, spent on sound files cutting and cleaning, after HTS voice training, we have really good and professionally-sounding result, ready to enter the public testing phase, and hopefully in two weeks or so ready to join the RHVoice.
Sorry for this personal story, but I believe that it can help.
To summarize: Try to build some team, able to do most parts of the work, and if you'll be  facing some blocking problems caused by RHVoice-related details, I'm sure that somebody here will help you by giving some advice.
The chances that somebody not using your language will be able and willing to implement it in whole just for fun, are rather small in my opinion, but who knows.
Most important: don't resign, and try to look for other people interested i the idea.
Have you any institution working for the blind in your country? Maybe it could be a good starting point in searching the other people to team.
And answering your last questions, all needed packages to install [are listed in the tutorial.](https://github.com/RHVoice/RHVoice/wiki/Creating-a-new-voice-for-RHVoice.)

[More about Macedonian in RHVoice.](https://louderpages.org/a-tool-that-will-help-over-10-thousand-blind-people/)

**Comment by king-dahmanus** [comment]
thanks for your comment.
There's not any institution for the blind in my country, and there are no arabic developers which can help. 2, I've met a developer, but when he didn't find the documentation he got frostrated and gave up, the foma reference is kinda way too advanced for me to learn from the beginning. There's not really a good way to make a voice unless I collabprate with someone like louderpages who knows what they're doing, cause there're not any devs who could do this as I said

**Comment by king-dahmanus** [comment]
actually there is a festival voice for arabic at https://github.com/linuxscout/festival-tts-arabic-voices if that'll help speed things up a litle

**Comment by king-dahmanus** [comment]
and it's not sapified. I think its developers discontinued it for some reason, because they didn't reply to my issue which requests sapification

**Comment by louderpages** [comment]
grzezio, I hope we can talk sometime, off-line.

Meanwhile, I would like to explain how Louder Pages operates.

There are two ways to make a voice. In one way, a team of developers decides to attempt a new voice.  They do this in their free time, which might be limited if they have a job, are studying or have a family to look after. They obtain favors from other people - studios, etc.. And they have to hope that those favors last until the work is complete.  They do. not know if they will succeed, and how long it will take them.  If they succeed, they will decide that they cannot further improve the voice’s quality, and they will offer it to the world as a gift.  Or, it may never be completed. If it is completed, then because it is a gift, it should not be criticized by the users.  (Although it will be. ha ha)

The other way, is where someone comes to the developers and says: "I want a voice.  I must have it in three months’ time. It must sound like this. This voice is so important for the world that If you, the developers, will promise in.a legal contract to do this for me, then I will pay you this much money when you give me the voice, and I check that it is the correct quality.”  

Louder Pages likes to make voices both ways.  The work on Macedonian and Albanian was of the second kind. That way is expensive because of the contractual promise.  We must not fail - on time, or quality.  We need that money to make this happen. To make sure the work gets done on time and properly.  And so.,the developers must dedicate their time fully to the work.  We need to enter into contracts with linguists, studios and actors, talk with lawyers, manage logistics, progress reports, etc..

For Louder Pages to do the first kind of development, we do it in our spare time in which case it can take forever, and perhaps we all go a little crazy with no other aspects to our life.  Or, we find some way of experimenting as part of our normal working time. We have a research budget.

The thing about the second way, is that if someone comes with funding, we can recruit, train and pay new developers to work in agreed timescale. We think that this is an effective way of bringing voices to the world. 

Arabic is very tough, because of its writing system..But it is widely spoken.  And there are some wealthy governments.  We live in hope. :)

Best,
- LP

**Comment by king-dahmanus** [comment]
I can make the dataset as of the time of this writing, I thankfully have a lot of freetime. Just instruct me, and I'll do half of the over all work as well as I can. Just tell me what to speak, how to organize the files and I'll do it as best as god helps me do

**Comment by king-dahmanus** [comment]
as for the diacritizer, I definitely recommend shakkelha https://github.com/AliOsm/shakkelha

**Comment by king-dahmanus** [comment]
bump, please don't forget this discussion. :smiles:

**Comment by king-dahmanus** [comment]
HI THERE. wHAT'S GOING ON?

**Comment by zstanecic** [comment]
What is going on with what?

Be more precise…

**Reply by louderpages** [reply]
We, here,  are all doing work on other voices, or in programming new features in RHVoice.  Or, doing other work elsewhere.

You kindly offer to do half the work in building an Arabic voice, but no one here has either the desire or the time available to do the other half.

You were advised quite well before, to go and find other people who may wish to help you. 

Good luck in your search.

**Comment by king-dahmanus** [comment]
e our discussion?we continu

**Comment by king-dahmanus** [comment]
@louderpages I'm sorry, I couldn't find anybody who wants to help make such
a thing

**Comment by mush42** [comment]
Hi @king-dahmanus @louderpages 

I'm developing a neural voice for Arabic.

First, I contributed Arabic language support to [Piper](https://rhasspy.github.io/piper-samples/), although the diacritizer used isn't the best.

Then I started working on a high-quality neural diacritizer, my work culminated in [Libtashkeel](https://github.com/mush42/libtashkeel/).
The open-source version is 8 MB for the model + 6 MB for ONNXRuntime. The proprietary  version is only 2 MB for the model and implemented directly in Rust. You can try out the FOSS version [here](https://mush42.github.io/libtashkeel/).

Although my main focus is on-device neural TTS, I might contribute to RHVoice if I ever have the time budget to do so.

Best
Musharraf

**Comment by zstanecic** [comment]
Hi @mush42 
If you want to start with this,
See the documentation for foma, this is a language in which language modules for RHVoice are written.
It is good to see the other languages as example. 
This is the first step.

**Comment by mush42** [comment]
Hi @zstanecic 

I'll take a look. As I said, my time budget is very limited these days.

Best

**Comment by king-dahmanus** [comment]
@louderpages greetings. I'm not sure whether to apologize for replying too late, for summoning you back here, or bringing back the topic. I had asked you about your contact info but you didn't answer. I'm chatting with Kimi about Arabic fst rules and I have no idea whether it's halucinating or not lol. Does rhvoice work fine with using another library like Libtashkeel developed by @mush42 and then passing on that processed text to Rhvoice, while the transducer has a bunch of rules for mood endings and grammatical cases? Does the transducer affect Rhvoice's intonation when the voice is trained? As in, does the training process corelate the voice info with the transducer, see how each grammatical case is intoned, then teaches it to the model so it isn't all monotonic or prone to excessive wrong intonation?
